api-server/scanner-lib/src/blockchain_state/mod.rs

```diff
@@ -91,3 +128,162 @@ impl<S: ApiServerStorage + Send + Sync> LocalBlockchainState for BlockchainState
+                                                        address.get(),
+                                                        new_amount,
+                                                        block_height,
+                                                    )
+                                                    .await?;
+                                            }
+                                        }
+                                    }
+                                    Destination::AnyoneCanSpend
+                                    | Destination::ClassicMultisig(_)
+                                    | Destination::ScriptHash(_) => {}
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    Ok(())
+}
+
+async fn update_balances_from_outputs<T: ApiServerStorageWrite>(
+    chain_config: Arc<ChainConfig>,
+    db_tx: &mut T,
+    block_height: BlockHeight,
+    outputs: &[TxOutput],
+) -> Result<(), ApiServerStorageError> {
+    for output in outputs {
+        match output {
+            TxOutput::Burn(_)
+            | TxOutput::CreateDelegationId(_, _)
+            | TxOutput::CreateStakePool(_, _)
+            | TxOutput::DataDeposit(_)
+            | TxOutput::DelegateStaking(_, _)
+            | TxOutput::IssueFungibleToken(_)
+            | TxOutput::IssueNft(_, _, _)
+            | TxOutput::ProduceBlockFromStake(_, _) => {}
+            TxOutput::Transfer(output_value, destination)
+            | TxOutput::LockThenTransfer(output_value, destination, _) => {
+                match destination {
+                    Destination::PublicKey(_) | Destination::Address(_) => {
+                        let address = Address::<Destination>::new(&chain_config, destination)
+                            .map_err(|_| {
+                                ApiServerStorageError::DeserializationError(
+                                    "Unable to encode destination".to_string(),
+                                )
+                            })?;
+
+                        match output_value {
+                            OutputValue::TokenV0(_) | OutputValue::TokenV1(_, _) => {
+                                // TODO
+                            }
+                            OutputValue::Coin(amount) => {
+                                let current_balance = db_tx
+                                    .get_address_balance(address.get())
+                                    .await?
+                                    .unwrap_or(Amount::ZERO);
+
+                                let new_amount = current_balance
+                                    .add(*amount)
+                                    .expect("Balance should not overflow");
+
+                                db_tx
+                                    .set_address_balance_at_height(
```

Comment:
I've been thinking about this, and I think this is wrong. If you have one transaction that has multiple outputs paying to the same destination, one will overwrite the other, leading to an invalid balance.

The correct way is to create an object that you keep updating with everything, and then write that in the database after iterating over all outputs.

Please write a test that proves this before fixing it; so something like TDD to ensure the problem is fixed.

Reply:
Ah you're right. Yep, TDD for this is good to prove it was incorrect to begin with.

Reply:
Interesting. I wrote the test and it passes without any fixes. Confused, I checked the logic and it seems to be correct - it loops over each output within the transaction, and gets the previous balance before updating to the new balance (i.e per output, not per transaction).